### ISSUE TITLE
Show average time to reply in profile

### ISSUE THREAD

**Issue by simison** [Trustroots/trustroots#420]
Include only threads:
- from the past 6 months (moving window)
- where user didn't initiate the message thread

Show some sort of fuzzy indication "replies within a day" etc

Value certainly requires some sort of caching.

The reasoning for this features is that 
1. people would tend to write to users who have better reply time
2. people would want to keep their reply time short publicly
3. highlights unresponsive users

Related to _show reply rate in profile_ (#421)

---

It could look like this:

```html
<div class="profile-sidebar-section text-muted">
  Reply rate 90%.<br>
  Replies within a day.
</div>
```

![image](https://cloud.githubusercontent.com/assets/87168/20387253/e2ca2b5c-acc7-11e6-875c-a086e0d56b6d.png)

**Comment by mrkvon** [comment]
Here is an attempt to mongoDB query, which returns amount of non-replied threads by user, amount of replied threads by user and average reply time. It works, the questions are: Does it work _correctly_? Can it be _simplified_? It fetches all the messages of a given user and then transforms them.

Detail:
- match all messages in which user is involved
- sort by `created`
- group by the thread and note who is the sender (`fromMe` boolean)
- match only threads started less than 180 days ago and started by the other user
- unwind the messages and then group the sent vs. received messages separately
- from each thread keep only the oldest message received and sent
- put the thread back together and count replyTime when available
- group, count amount of unreplied & replied threads and average replyTime

```
db.messages.aggregate([
{
  $match: {
    $or: [
      {userFrom: ObjectId("57f6ffc99838fb3ecdc94047")},
      {userTo: ObjectId("57f6ffc99838fb3ecdc94047")}
    ]
  }
},
{
  $sort: {created: 1}
},
{
  $group: {
    _id: {$cond: [
      {$eq: ["$userFrom", ObjectId("57f6ffc99838fb3ecdc94047")]},
      "$userTo",
      "$userFrom"
    ]},
    messages: {
      $push: {
        fromMe: {$cond: [
          {$eq: ["$userFrom", ObjectId("57f6ffc99838fb3ecdc94047")]},
          true,
          false
        ]},
        created: "$created"
      }
    }
  }
},
{
  $match: {
    "messages.0.fromMe": false,
    "messages.0.created": {$gte: new Date(Date.now()-180*24*3600*1000)}
  }
},
{
  $project: {
    "messages": 1,
    "id": "$_id",
    "_id": 0
  }
},
{
  $unwind: "$messages"
},
{
  $sort: {created: 1}
},
{
  $group: {
    _id: {id: "$id", "fromMe": "$messages.fromMe"},
    messages: {$push: "$messages.created"}
  }
},
{
  $project: {
    id: "$_id.id",
    _id: 0,
    fromMe: "$_id.fromMe",
    created: {$arrayElemAt: ["$messages", 0]}
  }
},
{
  $sort: {created: 1}
},
{
  $group: {
    _id: "$id",
    messages: {
      $push: "$created"
    }
  }
},
{
  $project: {
    _id: 0,
    replied: {$size: "$messages"},
    replyTime: {$cond: [
      {$eq: [{$size: "$messages"}, 2]},
      {$subtract: [
        {$arrayElemAt: ["$messages", 1]},
        {$arrayElemAt: ["$messages", 0]}
      ]},
      0
    ]}
  }
},
{
  $group: {
    _id: "$replied",
    replyTime: {$avg: "$replyTime"},
    count: {$sum: 1}
  }
}
]);
```

**Comment by mrkvon** [comment]
Where should the values appear?

- profile?
- offer?
- somewhere around messages?
- ?

**Comment by simison** [comment]
Just profile sidebar now at first first.

**Comment by simison** [comment]
It could be right above "Member since" block at the sidebar.

PS. For date stuff you can use Moment which is already available; `moment().subtract(6, 'months')` http://momentjs.com/docs/#/manipulating/subtract/ — a bit easier to read.

**Comment by simison** [comment]
From Workaway:

![image](https://cloud.githubusercontent.com/assets/87168/19219608/3a1936c6-8e20-11e6-9c04-920fba9cf14f.png)

**Comment by mrkvon** [comment]
Based on a discussion with @chmac and @simison here is a proposed way to implement (hopefully better) reply statistics:

a new model/mongo collection ThreadStat to keep track of the replied and non-replied Threads:
```js
{
  userFrom: UserId, // of the first message sender
  userTo: UserId, // of the first message receiver
  firstMessageCreated: Date,
  firstMessageLength: Number,
  firstReplyCreated: Date, // default null
  firstReplyLength: Number, // default null
  firstReplyTime: Number, // (ms) = firstReplyCreated - firstMessageCreated
  // ??? messageCount: count of all messages in the thread
}
```
another 2 fields with userIds in alphabetical order can be added, to ensure thread uniqueness with [multifield index](http://stackoverflow.com/questions/12573753/creating-multifield-indexes-in-mongoose-mongodb)

The following methods on the schema will be needed:
- find a ThreadStat when two users provided
- create new ThreadStat when first message is sent ([resource](http://stackoverflow.com/questions/16358857/mongodb-atomic-findorcreate-findone-insert-if-nonexistent-but-do-not-update))
- when firstReply is sent, update firstReplyCreated and firstReplyLength and firstReplyTime
- maybe [increment](https://docs.mongodb.com/manual/reference/operator/update/inc/) messageCount on every message sent (probably useless now)




_to be continued..._
_critics very welcome_

**Comment by chmac** [comment]
@mrkvon Schema looks good. I was thinking "first message", "second message", but you're absolutely right, it's really "first message" and "first reply". We're really thinking about first contact to first reply times at the moment.

I don't think we need to add `messageCount` at this point. I guess we can think about that later if it might be useful.

It might be clearer to call the first 2 fields `firstMessageUserFrom` and `firstMessageUserTo`. Leaves less room for confusion I think.

If we're calling it `ThreadStats`, then I'd suggest we also add the `threadId` and link the creation / update of these to the `Threads` collection rather than the messages collection. I can imagine scenarios in the future where we could have multiple threads for the same 2 users, so I'd suggest we avoid adding the unique constraint at this time on this collection.

If we think about messages -> threads -> thread stats, I think that's easier to manage than pushing from messages into 2 separate collections.

@mrkvon Thanks for your continued work on this.

**Comment by mrkvon** [comment]
implemented with #436 and #448
